### DISCUSSION TITLE
How to distribute a  new vehicle ( truck trailer)

### DISCUSSION THREAD

**Discussion by frankeng** [carla-simulator/carla#5476] [General]
Hi  I have build a working truck trailer combination in Carla and I like to distribute it.
I this not working perfectly and I like to have some help from people with more Carla and Unreal experience.
So can someone help me on how to distribute it in a correct way so that we can improve it together.
I tried the package methode ( make package ARGS="--packages=Package1,Package2") but I do not know what to fill in for the Pakage1 and when I run it I get 5 GB zip.

Here is a short impression of what I made. https://www.youtube.com/watch?v=sq825QASvZ4

**Comment by granadav** [comment]
It's great!!

**Comment by KevinLADLee** [comment]
That's impressive!

**Comment by hashim19** [comment]
Hi @frankeng, were you able to find out how we can distribute a custom vehicle to a docker version of carla package?

**Comment by frankeng** [comment]
No, Not with a docker version, I am now training to make a zip file which
you can unpack in the carla folder of the editor version.
I am training it with a truck made on a window version and unpacking it on
a linux version. but until now the linux version crashed so now and then.
I hope I will have the complete zip file next week. but at the moment I am
very busy.

Op vr 15 jul. 2022 om 18:34 schreef Hashim Ali ***@***.***>:

**Reply by hashim19** [reply]
From what I have understood so far, once you have the zipped package, you can import it in the editor version or the docker version both using the ./Import.sh script. Can you describe how you are creating the standalone package for your truck? I can see [this](https://carla.readthedocs.io/en/0.9.13/tuto_A_create_standalone/) tutorial but I am not sure what package1 and package2 is here? (the same problem as yours)

**Comment by DanielAtt2000** [comment]
@frankeng That looks very cool, would you be so kind in sharing your work? I have a school project I am working on and wish to use a truck and trailer combination. Maybe I can help with the distribution of your work/ package generation.

**Comment by DanielAtt2000** [comment]
@frankeng Any update on this?

**Comment by ikerpf** [comment]
Hi, @frankeng ! I am also interested into using and trying to improve the package. Is the package available anywhere to download it?

**Comment by frankeng** [comment]
Hi all I finally made a repository in github.
Please feel free to use and especially improve it.
Read the readme.md file for instructions, at the moment my carla is not working so I could not test if the instructions are complete and correct. But let me know if you have some questions:
https://github.com/frankeng/CarlaSemiTruckTrailer

**Reply by ikerpf** [reply]
Hi, Frank!

First of all, thanks for sharing! I could test the code in my computer, and the blueprints worked great, but coupling does not work for me. I tried both running the script and spawning the actors manually so close, but nothing happened. Maybe I did something wrong in the import proccess, so I have some questions about the coupling proccedure.

Does the coupling work for you? Or is there any keyboard shortcut to be pressed? Or is this one of those functionalities that Carla has lost and has to be fixed?

Just to check the version compatibility, I am using Unreal 4.26 and Carla 9.13

**Comment by frankeng** [comment]
Hi,
 it should just couple when the trailer is near to the truck, There is a
box on the the trailer if the truck gets into it it will couple ( or the
orherway arround).
Could it be that the name in the vehiclearry is not the same as in the blue
print ( I am just guessing now because at the moment I have not time to
look into it and mij carla is not working anymore.
Did you also try the traffic generator script?

Frank

Op ma 17 okt. 2022 om 10:20 schreef ikerpf ***@***.***>:

**Reply by ikerpf** [reply]
Ok, so I do not know why they do not couple in my device. I will keep looking at it. I had to write 'DAFxf' in lowercase ('dafxf') in the Python script for manual control (line 247) and call the script that generates traffic with --filterv instead of --filter, but in overall everything works fine (with the exception of the coupling, that remains imposible for me yet)

**Reply by frankeng** [reply]
@ikerpf 
Did you also tried to run 
 python3 manual_controlSemiTrailer.py
 Without the filter option? Then the truck and trailer are directly coupled by the python script. Beware that it truck is stuck in the beginning so it takes some seconds before it starts driving when you push the throttle.
Frank

**Reply by frankeng** [reply]
@ikerpf 
An update if you want to couple a trailer yourself (Manually), you should ad trailer in the editor on the road. And then run the original python script with the option -- filter dafxf
python manual_control.py -- filter dafxf.
If you then drive backwards to the trail it will couple to the trailer. It is not perfect yet but it should work.

**Reply by ikerpf** [reply]
Hi, @frankeng :

I tried both options:

1. Running python3 manual_controlSemiTrailer.py they spawn so close, but uncoupled. Then, when I drive backwards to the trailer, the head pushes the trailer, but they do not couple each other.
2. With the --filter dafxf option, it still happens the same.

I am uploading a video of the behavioral as you can see what happens: 
[Screencast from 24-10-22 08:54:12.webm](https://user-images.githubusercontent.com/65219611/197466686-a59ca4db-d155-4911-b09c-290d4f284b95.webm)

Do you think it can be because I am running in Linux instead of Windows? Or maybe are you using any add-on in Carla or Unreal that supports this coupling that I am not using and it is required?

**Reply by frankeng** [reply]
Hi @ikerpf 
That is strange I would think that it has nothing to with windows or linux, but I do not really know.
I more expecting that the files on github are not complete.
The coupling is done in the blueprints of the tractor and trailer. 
I will try to look at it this evening.

**Reply by frankeng** [reply]
@ikerpf 
Could you try this: just put a trailer in you world manually in the editor like this:
![image](https://user-images.githubusercontent.com/23728090/197488352-50ea165e-a2bd-43df-84bd-988bb012a247.png)
Then run the original `python manual_control.py --filter dafxf`
And drive the truck backwards to the trailer .
Let me know if this will work.
Frank

**Reply by ikerpf** [reply]
Yep, already done, but still the same. I do not know if my Carla is not letting Unreal execute the code inside the Blueprints or something, because I am looking the code inside the blueprints and it looks fine.

How do you execute the server? From the Unreal Editor, the compiled one, with make package or from an script?

**Reply by frankeng** [reply]
I run:  
`make launch` to start unreal.
press the play button in unreal.
And the start `python manual_control.py --filter dafxf`

**Reply by DanielAtt2000** [reply]
Hi @frankeng 

First all of thank you for sharing your work.

I think I have the same issue as above,

When running python3 manual_controlSemiTrailer.py, they are spawned next to each other but didnt actually couple. I also tried using manually putting it in Unity and driving the truck but still no luck.

Did you use any specific documentation when building this truck maybe I can try to find which bit is missing?
Also, maybe running a git diff can outline which files are different and therefore missing this repo

Thank you for your work again.

**Reply by frankeng** [reply]
HI @DanielAtt2000,
 I uses these videos:
https://www.youtube.com/watch?v=JwJplj92QoU&t=3414s
https://www.youtube.com/watch?v=mHgCuJc_Zh0
https://www.youtube.com/watch?v=c5IkFuJNMXE&t=12s
https://www.youtube.com/watch?v=mJufrK7RkeI&t=599s

I you found something please let me know, I am very curious what the problems is

Greetings
Frank

**Reply by DanielAtt2000** [reply]
Hi @frankeng 

I think the issue is because of the truck. There doesn't seem to be anything related to the hitch, as shown below.

![image](https://user-images.githubusercontent.com/59541958/198350811-9f926370-ca3c-4b42-8adf-60e291f9c00a.png)

On the other hand the trailer has hitch object.

![image](https://user-images.githubusercontent.com/59541958/198350939-ad01d83e-70e9-46e8-91d8-4103e9628e35.png)

Would you be able to check if you have another version of the truck with the correct components attached?
This would be very helpful for me 

Thank you for your help and time.
Daniel

**Reply by frankeng** [reply]
@DanielAtt2000,
That is helpfull I will check.
Frank

**Reply by frankeng** [reply]
@DanielAtt2000 
 I neither have ? I will look into it tomorrow.
![image](https://user-images.githubusercontent.com/23728090/198373835-9c65ab29-05eb-4ecd-8b42-cbeeef2c180c.png)

**Reply by DanielAtt2000** [reply]
@frankeng 

Maybe some other hints that will help find the problem.

Do you have this message when opening the EventGraph of the DAFTruck

![image](https://user-images.githubusercontent.com/59541958/198386615-94989f51-9efe-4961-b248-470100df9597.png)

Another possible hint, in _Couple The Trailer At Begin Play_ when you double click on _Mesh_ does error in the bottom right come up?

![image](https://user-images.githubusercontent.com/59541958/198386834-781e4de0-949a-4827-b16c-6ca723fc973b.png)

What if you do a git diff with the current carla version, do this show any other files which are different?
I will continue looking to maybe find the problem together.

Thank,
Daniel

**Reply by DanielAtt2000** [reply]
Did you create the Blueprint Class _Interactive Master?_ https://youtu.be/c5IkFuJNMXE?t=1789s

**Reply by DanielAtt2000** [reply]
@frankeng @ikerpf 

I got it working. The issue was that ticks where not enabled for the trailer.

You need to include a Set Actor Tick Enabled function on BP_Trailer. As shown below:

![image](https://user-images.githubusercontent.com/59541958/198662088-b2202206-efe5-4f9a-8b2a-9a391c22378a.png)

I will make a PR to your repo @frankeng.

**Reply by frankeng** [reply]
Hi @DanielAtt2000
 Good that you found it, I was already installing a new Carla to do a diff, but I can stop that now.

Good work!! I will look if i have the Set Actor Tick Enabled.
Could you push it to the repository?

Frank

**Reply by DanielAtt2000** [reply]
I did a pull request. If you can approve that that would be perfect!

**Reply by frankeng** [reply]
Hi @DanielAtt2000 
Strange I Did not have the Set Actor Tick Enabled function, but it still worked on my site?
I did read your message completely, I now saw that you already did a push, I will test it an than merge it to the main
![image](https://user-images.githubusercontent.com/23728090/198827281-c1f7da54-2403-4538-bf84-8e8bda4c587e.png)

**Reply by frankeng** [reply]
@DanielAtt2000 
I replied several time on your good news but I do not see the replies in this thread.
But again :-) great news. I already merged your fix, thanx.
I did not have the "Set Actor Tick Enabled function" in my blueprint but is was still working on my side. With the "Set Actor Tick Enabled function" it is also working.

 I have another question about git, I think you solved the problem. At the moment I do not have the code for the truck trailer coupled to this github repository, because I do not know how to do this. (combine the carla repository with the mine reposity)
I tried some thing with submodule but I could not figure out how to do it.
Can you tell me how you did it?
Frank

**Reply by DanielAtt2000** [reply]
If you want to have the code you changed and the original carla repo in the same place i would suggest doing something like this. 

https://opensource.com/article/19/7/create-pull-request-github

Then you can have something like frankeng/carla with a branch truck and trailer. Which has all of carla and your modifications

**Reply by ikerpf** [reply]
Amazing, guys! I just tried the new push and it worked great for me. I will try to help and solve the opened issues with it :)

**Reply by frankeng** [reply]
@ikerpf @DanielAtt2000 ,
You both run carla under Linux, right?

**Reply by DanielAtt2000** [reply]
I work using both. Though i haven't tried it on Linux yet.

**Reply by ikerpf** [reply]
Yep, I work in Linux and I can say it is absolutely compatible.

**Comment by frankeng** [comment]
I will try to look at it on friday

Op ma 17 okt. 2022 om 16:37 schreef ikerpf ***@***.***>:

**Comment by sojuzzoa** [comment]
Great work! Thanks for sharing your work!

**Comment by frankeng** [comment]
You can download a  zip file from git hub
[image: image.png]

Op di 31 mei 2022 om 13:36 schreef Sanju-Valeo ***@***.***>:

**Comment by frankeng** [comment]
HI @DanielAtt2000,
 I uses these videos:
https://www.youtube.com/watch?v=JwJplj92QoU&t=3414s
https://www.youtube.com/watch?v=mHgCuJc_Zh0
https://www.youtube.com/watch?v=c5IkFuJNMXE&t=12s
https://www.youtube.com/watch?v=mJufrK7RkeI&t=599s

I you found something please let me know, I am very curious what the
problems is

Greetings
Frank

Op wo 26 okt. 2022 om 12:10 schreef DanielAtt2000 ***@***.***

**Comment by DanielAtt2000** [comment]
I think you may have a tick box selected on the right when in the trailer
blue print (to enable ticking from there)

**Comment by frankeng** [comment]
ok thank you

Op do 3 nov. 2022 om 09:53 schreef ikerpf ***@***.***>:

**Comment by Sarang-Jokhio** [comment]
Hello,

Thank you for this great project. I am currently working with the truck and trailer setup provided here and would like to ask:

Is it possible to combine the truck and trailer into a single vehicle entity (e.g., as a unified blueprint or actor) that can be used just like other standard vehicles in CARLA and ScenarioRunner?

Specifically, I am interested in:

Treating the truck-trailer combo as one spawnable vehicle in ScenarioRunner (VehicleConfiguration in JSON or OpenSCENARIO)

Ensuring that control commands (throttle, brake, steer) apply to the whole unit without needing to manage trailer physics separately

Being able to assign it a behavior or route using AutoPilot or ScenarioRunner maneuvers, like any default CARLA vehicle

If this is feasible, could you advise on the best approach to do this in Unreal Engine and/or through code? Should the trailer be physically constrained and spawned within the truck blueprint, or is there a better way to encapsulate the two as a single mobility entity?

Thanks in advance!

**Reply by licongfei001** [reply]
Hello sir! Do you encounter that npc car would crash to the trailer? If only truck, it doesn't appear.

**Reply by DanielAtt2000** [reply]
@Sarang-Jokhio I am quite confident the truck-trailer vehicle can be combined into one unit. I believe that it wasn't implemented like that to preserve the individual units so they can be spawned individually. 

Having a quick think about it, I believe the following changes should be made
1) Updating the mesh and other visual aspects with the truck-trailer combined.
2) Updating the blueprint to remove the logic of the hitch which needs to be automatically attached on spawning of the vehicle.


The videos mentioned by @frankeng in https://github.com/carla-simulator/carla/discussions/5476#discussioncomment-3977824 are a really good starting point to see all the required steps.

Hope this helps and goodluck

**Comment by abhijeetbehera97** [comment]
Hi, 

You can find a new version (realistic, validated and compiled in higher Ubuntu) of the tractor-semitrailer in Carla. This is the continuation of the nice work done by @frankeng and @DanielAtt2000. 

Repository: https://github.com/abhijeetbehera97/Carla-Tractor-Semitrailer

**Comment by sunup494-pixel** [comment]
"Hi, I'd like to ask: Can custom vehicles in CARLA be packaged (as a standalone package) just like custom maps, and then unpacked and used directly in the pre-compiled (packaged binary) version of CARLA???"
